lib/torrent.js

```diff
@@ -1302,23 +1306,28 @@ export default class Torrent extends EventEmitter {
    * Garbage collect selections with respect to the store's current state.
    */
   _gcSelections () {
-    for (let i = 0; i < this._selections.length; ++i) {
-      const s = this._selections[i]
+    for (const s of this._selections) {
       const oldOffset = s.offset
 
       // check for newly downloaded pieces in selection
       while (this.bitfield.get(s.from + s.offset) && s.from + s.offset < s.to) {
         s.offset += 1
       }
 
-      if (oldOffset !== s.offset) s.notify()
+      if (oldOffset !== s.offset) {
+        for (const { notify } of this._selections.getNotificationsByRange({ from: s.from, to: s.from + s.offset })) {
+          notify()
+        }
+      }
       if (s.to !== s.from + s.offset) continue
       if (!this.bitfield.get(s.from + s.offset)) continue
 
-      this._selections.splice(i, 1) // remove fully downloaded selection
-      i -= 1 // decrement i to offset splice
-
-      s.notify()
+      s.remove() // remove fully downloaded selection
+      // remove notifications when we remove selection.
+      for (const notification of this._selections.getNotificationsByRange(s)) {
```

Comment:
and shouldnt this be notificationsInside?

the issue I'm seeing is that notifications which are touching, but arent containing a range, will get removed too, meaning they won't emit for other bigger ranges?